src/puppetlabs/puppetdb/catalogs.clj

```diff
@@ -367,13 +367,32 @@
            (throw (IllegalArgumentException. (format "Catalog must be specified as a string or a map, not '%s'" (class catalog)))))))
 
 (defmethod parse-catalog String
-  [catalog version]
+  [catalog version received-time]
   {:pre   [(string? catalog)]
    :post  [(map? %)]}
-  (parse-catalog (json/parse-string catalog true) version))
+  (parse-catalog (json/parse-string catalog true) version received-time))
+
+(defmethod parse-catalog 4
+  [catalog version received-time]
+  {:pre [(map? catalog)
+         (number? version)]
+   :post [(map? %)]}
+  (-> catalog
+      (assoc :producer_timestamp received-time)
+      (parse-catalog 5 received-time)))
+
+(defmethod parse-catalog 5
+  [catalog version received-time]
+  {:pre [(map? catalog)
+         (number? version)]
+   :post [(map? %)]}
+  (-> catalog
+      (set/rename-keys {:name :certname})
+      utils/dash->underscore-keys
```

Comment:
Oddly enough, this isn't a problem as utils/dash->underscore-keys only modifies the top level keys in the map. So although parameters isn't a problem, there's a lot of stuff it should be converting that it is not. I'm going to have to switch to using zippers here.